delivery.go

```diff
@@ -1,70 +1,122 @@
 package rmq
 
 import (
+	"context"
 	"fmt"
+	"time"
 )
 
 type Delivery interface {
 	Payload() string
-	Ack() bool
-	Reject() bool
-	Push() bool
+
+	Ack() error
+	Reject() error
+	Push() error
 }
 
-type wrapDelivery struct {
+type redisDelivery struct {
+	ctx         context.Context
 	payload     string
 	unackedKey  string
 	rejectedKey string
 	pushKey     string
 	redisClient RedisClient
+	errChan     chan<- error
 }
 
-func newDelivery(payload, unackedKey, rejectedKey, pushKey string, redisClient RedisClient) *wrapDelivery {
-	return &wrapDelivery{
+func newDelivery(
+	ctx context.Context,
+	payload string,
+	unackedKey string,
+	rejectedKey string,
+	pushKey string,
+	redisClient RedisClient,
+	errChan chan<- error,
+) *redisDelivery {
+	return &redisDelivery{
+		ctx:         ctx,
 		payload:     payload,
 		unackedKey:  unackedKey,
 		rejectedKey: rejectedKey,
 		pushKey:     pushKey,
 		redisClient: redisClient,
+		errChan:     errChan,
 	}
 }
 
-func (delivery *wrapDelivery) String() string {
+func (delivery *redisDelivery) String() string {
 	return fmt.Sprintf("[%s %s]", delivery.payload, delivery.unackedKey)
 }
 
-func (delivery *wrapDelivery) Payload() string {
+func (delivery *redisDelivery) Payload() string {
 	return delivery.payload
 }
 
-func (delivery *wrapDelivery) Ack() bool {
-	// debug(fmt.Sprintf("delivery ack %s", delivery)) // COMMENTOUT
+// blocking versions of the functions below with the following behavior:
+// 1. return immediately if the operation succeeded or failed with ErrorNotFound
+// 2. in case of other redis errors, send them to the errors chan and retry after a sleep
+// 3. if redis errors occur after StopConsuming() has been called, ErrorConsumingStopped will be returned
+
+func (delivery *redisDelivery) Ack() error {
+	errorCount := 0
+	for {
+		count, err := delivery.redisClient.LRem(delivery.unackedKey, 1, delivery.payload)
+		if err == nil { // no redis error
+			if count == 0 {
+				return ErrorNotFound
+			}
+			return nil
+		}
+
+		// redis error
+
+		errorCount++
+
+		select { // try to add error to channel, but don't block
+		case delivery.errChan <- &DeliveryError{Delivery: delivery, RedisErr: err, Count: errorCount}:
+		default:
+		}
+
+		if err := delivery.ctx.Err(); err != nil {
+			return ErrorConsumingStopped
```

Comment:
the only way to exit the Ack() function in case of error is to stop the consumer, which might not be desired in some use cases. shall we have some `retry` configuration for ack and return the last error if it reaches the count?

Reply:
I also feel like an infinite loop, which only terminates on success or shutdown is quite severe.

But this is moderated by the fact that if a redis connection remains bad for long enough the heartbeat will exit and shutdown all the consumers.

For me having Acks etc. being linked to the heartbeat is good enough.

Reply:
Generally I was trying here to keep the concepts simple. If we just keep retrying indefinitely we don't need to be concerned about additional deliveries being consumed in an unstable redis state.

@NeoHuang: Can you share a use case where you think a retry limit might be beneficial? And what would you want to do after the retry limit has been reached and you got your error? Would you just want to continue working on deliveries which you are very likely to not be able to ack again? To me that seems more dangerous than not trying to consume anything until the connection is back.

Reply:
alright, I think about it again. since we have the heartbeat to stop the consumer, I think that might be fine